File: sac2asc.f
      Program sac2asc
      integer npts1,npts2,i,nptsz
      real dt1,dt2,dtz,Parr,Sarr,t,t1,t2,b,mz
      real y1(100000),y2(100000),yz(100000)
      character*50 fn1,fn2,fnz
      character*1  typ

      open(100,file='result')

      read(*,*)typ
      read(*,*)fnz
      read(*,*)fn1
      read(*,*)fn2

      call RSAC1(fnz,yz,nlen,beg,del,100000,nerr)
      call getnhv('NPTS',nptsz,nerr)
      call getfhv('DELTA',dtz,nerr)

      call RSAC1(fn1,y1,nlen,beg,del,100000,nerr)
      call getnhv('NPTS',npts1,nerr)
      call getfhv('DELTA',dt1,nerr)
      call getfhv('B',b,nerr)
      call getfhv('T1',t1,nerr)
      call getfhv('T2',t2,nerr)
      call getfhv('T3',t3,nerr)
      call getfhv('T4',t4,nerr)

      call RSAC1(fn2,y2,nlen,beg,del,100000,nerr)
      call getnhv('NPTS',npts2,nerr)
      call getfhv('DELTA',dt2,nerr)

      Parr=t1-b
      Sarr=t2-b

      if ((t4 - (-12345)) < 0.00000001)then
         t4 = b
      endif

c     rmean Z-comp.
      mz=sum(yz)/npts1
      do i = 1, npts1
         yz(i)=yz(i)-mz
      enddo

c      Output ascii
      open(100,file='data.asc',status='new')
      write(100,*)Parr,Sarr,typ
      do i = 1, npts1
           t=(i-1)*dt1
           if (t <= t3-b .and. t >= t4-b )then
               write(100,'(f10.5,2X,f10.5,2X,f10.5,2X,f10.5)')
     1             t,yz(i),y1(i),y2(i)
           endif
      enddo
      close(100)      
      stop
      end

File: sac2asc.py
class sac2asc:
    def __init__(self,sacdataz,sacdata1,sacdata2,zory):
        self.npts1 = int(sacdata1.npts)
        self.npts2 = int(sacdata2.npts)
        self.nptsz = int(sacdataz.npts)
        self.dt1 = float(sacdata1.delta)
        self.dt2 = float(sacdata2.delta)
        self.dtz = float(sacdataz.delta)
        try:
            self.t1 = float(sacdata1.t1)
            self.t2 = float(sacdata1.t2)
            self.t3 = float(sacdata1.t3)
            self.t4 = float(sacdata1.t4)
        except:
            print("plz define t1-t4")
        self.b = float(sacdata1.b)
        self.y1 = sacdata1.data
        self.y2 = sacdata2.data
        self.yz = sacdataz.data
        self.Parr = float(sacdata1.t1-sacdata1.b)
        self.Sarr = float(sacdata1.t2-sacdata1.b)
        self.zory = zory

        self.mz=sum(self.yz)/self.npts1 # yz -> mean
        for i in range(self.npts1):
            self.yz[i]=self.yz[i]-self.mz # normalize
        self.m1=sum(self.y1)/self.npts1
        for i in range(self.npts1):
            self.y1[i]=self.y1[i]-self.m1 # normalize
        self.m2=sum(self.y2)/self.npts1
        for i in range(self.npts1):
            self.y2[i]=self.y2[i]-self.m2 # normalize
            
    def __call__(self):
        with open("data.asc","w") as file: # !namex
            file.write(" ")
            file.write(str(round(self.Parr,4)))
            file.write("\t")
            file.write(str(round(self.Sarr,4)))
            file.write("\t")
            if self.zory == "z":
                file.write("9")
            elif self.zory == "y":
                file.write("0")
            file.write("\n") # !type
            for i in range(self.npts1):
                t = (i-1)*self.dt1
                if t<=self.t3-self.b and t>=self.t4-self.b:
                    file.write(" ")
                    file.write(str(round(t,4)))
                    file.write("   ")
                    file.write(str(round(self.yz[i],4)))
                    file.write("   ")
                    file.write(str(round(self.y1[i],4)))
                    file.write("   ")
                    file.write(str(round(self.y2[i],4)))
                    file.write("\n")